File: media/audiobooks.py
"""
media/audiobooks.py
-------------------
Audiobook-specific parsing, serialization, and DB logic for MOASYS-Vault.

Expected folder structure:
  <media_folder>/
    <Author Name>/                          ← single author
    <Author 1, Author 2>/                   ← multiple authors, comma-separated
    <Author 1, Author 2, and Author 3>/     ← multiple authors with "and"
      <Book Title>/
        01 - Chapter Name.m4b
        101 - Chapter Name.mp3   ← multi-disc (Book On CD)
"""

import os
import re
import json

# ─────────────────────────────────────────────
# Regex patterns
# ─────────────────────────────────────────────

# Matches standard chapter file stems: "01 - Chapter Name"
# Group 1 = chapter number (2 digits), Group 2 = chapter name
SINGLE_DISC_PATTERN = re.compile(r'^(\d{2})\s-\s(.+)$')

# Matches multi-disc chapter file stems: "101 - Chapter Name", "201 - Chapter Name"
# Group 1 = disc number, Group 2 = chapter number (2 digits), Group 3 = chapter name
MULTI_DISC_PATTERN = re.compile(r'^(\d+)(\d{2})\s-\s(.+)$')


# ─────────────────────────────────────────────
# Helpers
# ─────────────────────────────────────────────

def _is_audio(filename, media_config):
    """Return True if the file extension is in the configured audio_extensions list."""
    _, ext = os.path.splitext(filename)
    return ext.lower() in [e.lower() for e in media_config["audio_extensions"]]

def _get_primary_extensions(media_config):
    """
    Return the primary extension(s) as a lowercase list.
    config value must be a list e.g. [".m4b"] or [".m4b", ".mp3"].
    """
    return [e.lower() for e in media_config["primary_extension"]]

def _is_primary(filename, media_config):
    """Return True if the file extension matches any configured primary_extension."""
    _, ext = os.path.splitext(filename)
    return ext.lower() in _get_primary_extensions(media_config)

def _format_primary_exts(media_config):
    """
    Return a human-readable string of primary extensions for warning messages.
    e.g. [".m4b"] -> "Non-.M4B"
         [".m4b", ".mp3"] -> "Non-.M4B/.MP3"
    """
    exts = _get_primary_extensions(media_config)
    return "Non-" + "/".join(e.upper() for e in exts)

def _parse_authors(folder_name):
    """
    Parse an author folder name into a list of individual author names.
    Handles three formats:
      "J.R.R. Tolkien"                        -> ["J.R.R. Tolkien"]
      "Terry Pratchett, Neil Gaiman"           -> ["Terry Pratchett", "Neil Gaiman"]
      "Author 1, Author 2, and Author 3"       -> ["Author 1", "Author 2", "Author 3"]

    Strategy:
      1. Strip trailing " and " from the last segment
      2. Split on commas
      3. Strip whitespace from each name
    """
    # Remove " and " that may appear before the last author
    # e.g. "Author 1, Author 2, and Author 3" -> "Author 1, Author 2, Author 3"
    cleaned = re.sub(r',?\s+and\s+', ', ', folder_name, flags=re.IGNORECASE)
    authors = [a.strip() for a in cleaned.split(',') if a.strip()]
    return authors

def _parse_chapter_stem(stem):
    """
    Parse a chapter file stem into (disc_number, chapter_number, chapter_name).
    Handles both single-disc (01 - Name) and multi-disc (101 - Name) formats.

    Returns (disc, chapter, name) or None if no match.

    Single-disc treated as disc 1:
      "01 - The Shadow of the Past" -> (1, 1, "The Shadow of the Past")

    Multi-disc extracts disc from leading digit(s):
      "101 - Chapter Name" -> (1, 1, "Chapter Name")
      "302 - Chapter Name" -> (3, 2, "Chapter Name")
    """
    # Try multi-disc first — it's more specific
    m = MULTI_DISC_PATTERN.match(stem)
    if m:
        disc    = int(m.group(1))
        chapter = int(m.group(2))
        name    = m.group(3).strip()
        return disc, chapter, name

    # Fall back to single-disc
    m = SINGLE_DISC_PATTERN.match(stem)
    if m:
        chapter = int(m.group(1))
        name    = m.group(2).strip()
        return 1, chapter, name   # Treat as disc 1

    return None

def _make_book_key(title):
    """
    Unique key for a book — lowercased title only.
    Books are keyed by title since they're the top-level item in the output.
    The same book shouldn't appear under different author folders.
    """
    return title.lower()

def _find_chapter_gaps(chapter_numbers):
    """
    Given a list of chapter numbers for a single disc, return any missing numbers.
    Example: [1, 2, 4] -> [3]
    """
    if not chapter_numbers:
        return []
    full_range = set(range(min(chapter_numbers), max(chapter_numbers) + 1))
    return sorted(full_range - set(chapter_numbers))


# ─────────────────────────────────────────────
# Scanner (called by core/scanner.py)
# ─────────────────────────────────────────────

def scan_quality_folder(quality_path, folder_name, tag, media_config, warnings):
    """
    Walk one media folder (e.g. Audible/, Book On CD/) and return a dict of book records.
    Called once per media_folder entry by core/scanner.py.

    Returns:
    {
      book_key: {
        title, authors: list, chapter_count, media_type: set()
      }
    }

    Note: Unlike music, books are keyed by title only — not author + title.
    This is because a book is the primary searchable item, and authors
    are stored as an array field on the book record.
    """
    records = {}

    # Each subfolder inside the media folder should be an author folder
    for author_entry in os.scandir(quality_path):
        if not author_entry.is_dir():
            continue   # Skip loose files in the media folder

        author_folder = author_entry.name
        author_rel    = os.path.join(folder_name, author_folder)

        # Parse the folder name into a list of authors
        # e.g. "Terry Pratchett, Neil Gaiman" -> ["Terry Pratchett", "Neil Gaiman"]
        authors = _parse_authors(author_folder)

        # Read book folders inside this author folder
        try:
            book_entries = list(os.scandir(author_entry.path))
        except PermissionError:
            warnings.add(author_rel, "Permission denied reading author folder")
            continue

        for book_entry in book_entries:
            if not book_entry.is_dir():
                continue   # Skip loose files (e.g. author photo)

            book_title = book_entry.name
            book_rel   = os.path.join(author_rel, book_title)
            book_key   = _make_book_key(book_title)

            # Read chapter files inside this book folder
            try:
                all_files = list(os.scandir(book_entry.path))
            except PermissionError:
                warnings.add(book_rel, "Permission denied reading book folder")
                continue

            audio_files   = [f for f in all_files if _is_audio(f.name, media_config)]
            non_primary   = [f for f in audio_files if not _is_primary(f.name, media_config)]

            # Warning: no audio files at all in this book folder
            if not audio_files:
                warnings.add(book_rel, "No recognized audio files found in book folder")
                continue

            # Warning: non-primary audio files present
            for f in non_primary:
                _, ext = os.path.splitext(f.name)
                warnings.add(
                    os.path.join(book_rel, f.name),
                    f"{_format_primary_exts(media_config)} audio file — may need re-encoding",
                    extension=ext.lower()
                )

            # Track chapter numbers per disc for gap detection: { disc_num: [chapter_nums] }
            disc_chapters = {}
            chapter_count = 0

            for vf in audio_files:
                stem, _ = os.path.splitext(vf.name)
                parsed  = _parse_chapter_stem(stem)

                # Warning: file name doesn't match naming convention
                if not parsed:
                    warnings.add(
                        os.path.join(book_rel, vf.name),
                        "Chapter file name does not match naming convention — "
                        "expected: 01 - Chapter Name.ext or 101 - Chapter Name.ext (multi-disc)"
                    )
                    continue

                disc, chapter, _ = parsed
                chapter_count += 1

                # Group chapter numbers by disc for per-disc gap detection
                if disc not in disc_chapters:
                    disc_chapters[disc] = []
                disc_chapters[disc].append(chapter)

            # Warning: gaps in chapter numbers, checked per disc independently
            for disc_num, chapters in sorted(disc_chapters.items()):
                gaps = _find_chapter_gaps(chapters)
                if gaps:
                    gap_str  = ", ".join(f"Chapter {g:02d}" for g in gaps)
                    disc_str = f"Disc {disc_num}" if len(disc_chapters) > 1 else "Book"
                    warnings.add(
                        book_rel,
                        f"Potential missing chapters in {disc_str}: {gap_str}"
                    )

            # Add / merge book into records
            if book_key not in records:
                records[book_key] = {
                    "title":         book_title,
                    "authors":       authors,
                    "chapter_count": chapter_count,
                    "media_type":    set()
                }
            else:
                # Book already exists — update chapter count if higher
                records[book_key]["chapter_count"] = max(
                    records[book_key]["chapter_count"],
                    chapter_count
                )

            records[book_key]["media_type"].add(tag)

            # Warning: same book found in more than one media folder
            if len(records[book_key]["media_type"]) > 1:
                existing_tags = ", ".join(sorted(records[book_key]["media_type"]))
                warnings.add(
                    book_rel,
                    f"Duplicate book found in multiple media folders: {existing_tags}"
                )

    return records


# ─────────────────────────────────────────────
# Merge (called by core/scanner.py)
# ─────────────────────────────────────────────

def merge(existing, new_records):
    """
    Merge new_records into existing records across media folders.
    Books are keyed by title — if the same title appears in multiple
    media folders, media_type accumulates both tags.
    """
    for book_key, new_book in new_records.items():
        if book_key not in existing:
            existing[book_key] = new_book
        else:
            # Merge media_type sets and keep highest chapter count
            existing[book_key]["media_type"].update(new_book["media_type"])
            existing[book_key]["chapter_count"] = max(
                existing[book_key]["chapter_count"],
                new_book["chapter_count"]
            )


# ─────────────────────────────────────────────
# Quality order / media type order
# ─────────────────────────────────────────────

_MEDIA_TYPE_ORDER = []

def init_quality_order(media_folders):
    """Called by scan.py at startup — stores media folder tag order for media_type sorting."""
    global _MEDIA_TYPE_ORDER
    _MEDIA_TYPE_ORDER = [mf["tag"] for mf in media_folders]

def _order_media_type(media_type_set):
    """Sort media_type tags in config-defined order."""
    return [t for t in _MEDIA_TYPE_ORDER if t in media_type_set]


# ─────────────────────────────────────────────
# Serializer (called by core/scanner.py)
# ─────────────────────────────────────────────

def serialize(records):
    """
    Convert the internal records dict into a sorted list of dicts for JSON output.
    Books are sorted alphabetically by title.
    Authors list preserves the order parsed from the folder name.
    """
    return sorted(
        [
            {
                "title":         r["title"],
                "authors":       r["authors"],
                "chapter_count": r["chapter_count"],
                "media_type":    _order_media_type(r["media_type"])
            }
            for r in records.values()
        ],
        key=lambda x: x["title"].lower()
    )


# ─────────────────────────────────────────────
# DB writer (called by core/scanner.py)
# ─────────────────────────────────────────────

def write_db(conn, records):
    """
    Write audiobook records to SQLite.
    Single table — books are the top-level item.
    authors and media_type stored as JSON strings.
    """
    cur = conn.cursor()

    cur.execute("""
        CREATE TABLE IF NOT EXISTS books (
            id            INTEGER PRIMARY KEY AUTOINCREMENT,
            title         TEXT    NOT NULL,
            authors       TEXT    NOT NULL,   -- JSON array e.g. '["J.R.R. Tolkien"]'
            chapter_count INTEGER NOT NULL,
            media_type    TEXT    NOT NULL,   -- JSON array e.g. '["Audible"]'
            UNIQUE(title)
        )
    """)

    for r in records.values():
        authors_json    = json.dumps(r["authors"])
        media_type_json = json.dumps(_order_media_type(r["media_type"]))
        cur.execute("""
            INSERT INTO books (title, authors, chapter_count, media_type)
            VALUES (?, ?, ?, ?)
            ON CONFLICT(title) DO UPDATE SET
                authors       = excluded.authors,
                chapter_count = excluded.chapter_count,
                media_type    = excluded.media_type
        """, (r["title"], authors_json, r["chapter_count"], media_type_json))

File: src/media/audiobooks.ts
/**
 * media/audiobooks.ts
 * -------------------
 * Audiobook-specific parsing, serialization, and DB logic for MOASYS-Vault.
 *
 * Expected folder structure:
 *   <media_folder>/
 *     <Author Name>/                          ← single author
 *     <Author 1, Author 2>/                   ← multiple authors, comma-separated
 *     <Author 1, Author 2, and Author 3>/     ← multiple authors with "and"
 *       <Book Title>/
 *         01 - Chapter Name.m4b
 *         101 - Chapter Name.mp3   ← multi-disc (Book On CD)
 */

import fs from 'fs'
import path from 'path'

import {
  AudiobooksConfig,
  MediaFolder,
  BookRecord,
  BookOutput,
  WarningCollector,
  MediaModule,
} from '../core/types'

// ─────────────────────────────────────────────
// Regex patterns
// ─────────────────────────────────────────────

// Matches standard chapter file stems: "01 - Chapter Name"
// Group 1 = chapter number (2 digits), Group 2 = chapter name
const SINGLE_DISC_PATTERN = /^(\d{2})\s-\s(.+)$/

// Matches multi-disc chapter file stems: "101 - Chapter Name", "201 - Chapter Name"
// Group 1 = disc number, Group 2 = chapter number (2 digits), Group 3 = chapter name
const MULTI_DISC_PATTERN = /^(\d+)(\d{2})\s-\s(.+)$/

// ─────────────────────────────────────────────
// Helpers
// ─────────────────────────────────────────────

function isAudio(filename: string, config: AudiobooksConfig): boolean {
  const ext = path.extname(filename).toLowerCase()
  return config.audio_extensions.map(e => e.toLowerCase()).includes(ext)
}

function isPrimary(filename: string, config: AudiobooksConfig): boolean {
  const ext = path.extname(filename).toLowerCase()
  return config.primary_extension.map(e => e.toLowerCase()).includes(ext)
}

function formatPrimaryExts(config: AudiobooksConfig): string {
  return 'Non-' + config.primary_extension.map(e => e.toUpperCase()).join('/')
}

/**
 * Parse an author folder name into a list of individual author names.
 * Handles three formats:
 *   "J.R.R. Tolkien"                    -> ["J.R.R. Tolkien"]
 *   "Terry Pratchett, Neil Gaiman"       -> ["Terry Pratchett", "Neil Gaiman"]
 *   "Author 1, Author 2, and Author 3"  -> ["Author 1", "Author 2", "Author 3"]
 *
 * Strategy: strip " and " before the last author, then split on commas.
 */
function parseAuthors(folderName: string): string[] {
  // Remove " and " (with optional leading comma) before the last author
  const cleaned = folderName.replace(/,?\s+and\s+/gi, ', ')
  return cleaned
    .split(',')
    .map(a => a.trim())
    .filter(a => a.length > 0)
}

/**
 * Parse a chapter file stem into { disc, chapter, name } or null.
 * Single-disc files (01 - Name) are treated as disc 1.
 * Multi-disc files (101 - Name) extract the disc from leading digit(s).
 */
function parseChapterStem(stem: string): { disc: number; chapter: number; name: string } | null {
  // Try multi-disc first — it's more specific
  let m = MULTI_DISC_PATTERN.exec(stem)
  if (m) {
    return {
      disc: parseInt(m[1]!, 10),
      chapter: parseInt(m[2]!, 10),
      name: m[3]!.trim(),
    }
  }
  // Fall back to single-disc (treat as disc 1)
  m = SINGLE_DISC_PATTERN.exec(stem)
  if (m) {
    return { disc: 1, chapter: parseInt(m[1]!, 10), name: m[2]!.trim() }
  }
  return null
}

/**
 * Build a unique Map key for a book.
 * Keyed by title only — not author + title — because books are the
 * top-level item in the output and authors are stored as an array field.
 */
function makeBookKey(title: string): string {
  return title.toLowerCase()
}

/**
 * Find gaps in chapter numbers for a single disc.
 * Example: [1, 2, 4] -> [3]
 */
function findChapterGaps(numbers: number[]): number[] {
  if (numbers.length === 0) return []
  const min = Math.min(...numbers)
  const max = Math.max(...numbers)
  const gaps = []
  for (let i = min; i <= max; i++) {
    if (!numbers.includes(i)) gaps.push(i)
  }
  return gaps
}

// ─────────────────────────────────────────────
// Media type order
// ─────────────────────────────────────────────

let mediaTypeOrder: string[] = []

// ─────────────────────────────────────────────
// Media module implementation
// ─────────────────────────────────────────────

export const audiobooksModule: MediaModule<BookRecord, BookOutput, AudiobooksConfig> = {
  initTagOrder(mediaFolders: MediaFolder[]): void {
    mediaTypeOrder = mediaFolders.map(mf => mf.tag)
  },

  scanMediaFolder(
    folderPath: string,
    folderName: string,
    tag: string,
    config: AudiobooksConfig,
    warnings: WarningCollector
  ): Map<string, BookRecord> {
    const records = new Map<string, BookRecord>()

    // Each subfolder inside the media folder should be an author folder
    for (const authorEntry of fs.readdirSync(folderPath, {
      withFileTypes: true,
    })) {
      if (!authorEntry.isDirectory()) continue

      const authorPath = path.join(folderPath, authorEntry.name)
      const authorRel = path.join(folderName, authorEntry.name)

      // Parse the folder name into a list of authors
      // e.g. "Terry Pratchett, Neil Gaiman" -> ["Terry Pratchett", "Neil Gaiman"]
      const authors = parseAuthors(authorEntry.name)

      let bookEntries: fs.Dirent[]
      try {
        bookEntries = fs.readdirSync(authorPath, { withFileTypes: true })
      } catch {
        warnings.add(authorRel, 'Permission denied reading author folder')
        continue
      }

      for (const bookEntry of bookEntries) {
        if (!bookEntry.isDirectory()) continue // Skip loose files (e.g. author photo)

        const bookPath = path.join(authorPath, bookEntry.name)
        const bookRel = path.join(authorRel, bookEntry.name)
        const bookKey = makeBookKey(bookEntry.name)

        let allFiles: fs.Dirent[]
        try {
          allFiles = fs.readdirSync(bookPath, { withFileTypes: true })
        } catch {
          warnings.add(bookRel, 'Permission denied reading book folder')
          continue
        }

        const audioFiles = allFiles.filter(f => f.isFile() && isAudio(f.name, config))
        const nonPrimary = audioFiles.filter(f => !isPrimary(f.name, config))

        // Warning: no audio files at all in this book folder
        if (audioFiles.length === 0) {
          warnings.add(bookRel, 'No recognized audio files found in book folder')
          continue
        }

        // Warning: non-primary audio files present
        for (const f of nonPrimary) {
          const ext = path.extname(f.name).toLowerCase()
          warnings.add(
            path.join(bookRel, f.name),
            `${formatPrimaryExts(config)} audio file — may need re-encoding`,
            ext
          )
        }

        // Track chapter numbers per disc for gap detection: { discNum: [chapterNums] }
        const discChapters = new Map<number, number[]>()
        let chapterCount = 0

        for (const f of audioFiles) {
          const stem = path.basename(f.name, path.extname(f.name))
          const parsed = parseChapterStem(stem)

          // Warning: file name doesn't match naming convention
          if (!parsed) {
            warnings.add(
              path.join(bookRel, f.name),
              'Chapter file name does not match naming convention — ' +
                'expected: 01 - Chapter Name.ext or 101 - Chapter Name.ext (multi-disc)'
            )
            continue
          }

          const { disc, chapter } = parsed
          chapterCount++

          if (!discChapters.has(disc)) discChapters.set(disc, [])
          discChapters.get(disc)!.push(chapter)
        }

        // Warning: gaps in chapter numbers, checked per disc independently
        for (const [discNum, chapters] of [...discChapters.entries()].sort(([a], [b]) => a - b)) {
          const gaps = findChapterGaps(chapters)
          if (gaps.length > 0) {
            const gapStr = gaps.map(g => `Chapter ${String(g).padStart(2, '0')}`).join(', ')
            const discStr = discChapters.size > 1 ? `Disc ${discNum}` : 'Book'
            warnings.add(bookRel, `Potential missing chapters in ${discStr}: ${gapStr}`)
          }
        }

        // Add or merge book into records
        if (!records.has(bookKey)) {
          records.set(bookKey, {
            title: bookEntry.name,
            authors,
            chapter_count: chapterCount,
            media_type: new Set(),
          })
        } else {
          // Book already exists — update chapter count if higher
          records.get(bookKey)!.chapter_count = Math.max(
            records.get(bookKey)!.chapter_count,
            chapterCount
          )
        }

        records.get(bookKey)!.media_type.add(tag)

        // Warning: same book title found in more than one media folder
        if (records.get(bookKey)!.media_type.size > 1) {
          const existingTags = [...records.get(bookKey)!.media_type].sort().join(', ')
          warnings.add(bookRel, `Duplicate book found in multiple media folders: ${existingTags}`)
        }
      }
    }

    return records
  },

  /**
   * Merge book records across media folders.
   * Books are keyed by title — same title in multiple folders merges media_type.
   */
  merge(existing: Map<string, BookRecord>, incoming: Map<string, BookRecord>): void {
    for (const [bookKey, newBook] of incoming) {
      if (!existing.has(bookKey)) {
        existing.set(bookKey, newBook)
      } else {
        const existingBook = existing.get(bookKey)!
        for (const t of newBook.media_type) existingBook.media_type.add(t)
        existingBook.chapter_count = Math.max(existingBook.chapter_count, newBook.chapter_count)
      }
    }
  },

  /** Convert records to sorted array for JSON output. Books sorted alphabetically by title. */
  serialize(records: Map<string, BookRecord>): BookOutput[] {
    const orderMediaType = (mt: Set<string>) => mediaTypeOrder.filter(t => mt.has(t))

    return [...records.values()]
      .sort((a, b) => a.title.toLowerCase().localeCompare(b.title.toLowerCase()))
      .map(book => ({
        title: book.title,
        authors: book.authors,
        chapter_count: book.chapter_count,
        media_type: orderMediaType(book.media_type),
      }))
  },
}
